psu/psu.c

```diff
@@ -164,440 +93,513 @@ static inline void set_write_reg_cmd(unsigned char *b, uint32_t addr, uint8_t fo
-		fprintf(stderr, "Failed to receive HAB mode (n=%d)\n", rc);
-		return rc;
+	if (!(s->next.str.end - s->next.str.ptr == 1 && (*s->next.str.ptr == 'F' || *s->next.str.ptr == 'S'))) {
+		s->errstr = "Type F or S expected";
+		return SCRIPT_ERROR;
 	}
 
-	if((rc = hid_read(dev, b, INTERRUPT_SIZE)) < 0) {
-		fprintf(stderr, "Failed to receive status (n=%d)\n", rc);
-		return rc;
-	}
+	if (script_expect(s, script_tok_identifier, "Literal F or S expected") != SCRIPT_OK)
+		return SCRIPT_ERROR;
 
-	return rc;
-}
+	type = *s->token.str.ptr;
 
+	if (script_expect(s, script_tok_string, "String in quotes was expected") != SCRIPT_OK)
+		return SCRIPT_ERROR;
 
-int execute_line(char *line, size_t len, size_t lineno, hid_device **dev)
-{
-	int err = 0;
-	char *tok = strtok(line, " ");
-	size_t toklen = strlen(tok);
+	str = s->token.str;
 
-	if (tok[0] != '\n' && tok[0] != '#') { /* Skip empty lines and comments */
-		if (tok[toklen - 1] == '\n')
-			tok[--toklen] = '\0';
+	if (script_expect_opt(s, script_tok_integer, "Optional <offset> value was expected") == SCRIPT_OK)
+		offset = s->token.num;
 
-		fprintf(stderr, "Parsing %lu: '%s'\n", lineno, tok);
+	if (s->errstr)
+		return SCRIPT_ERROR;
 
-		if (!strcmp(tok, "WAIT")) {
-			err = wait_cmd(dev);
-		}
-		else if(!strcmp(tok, "WRITE_FILE")) {
-			err = write_file_cmd(*dev);
-		}
-		else if(!strcmp(tok, "WRITE_REGISTER")) {
-			err = write_reg_cmd(*dev);
-		}
-		else if(!strcmp(tok, "JUMP_ADDRESS")) {
-			err = jump_addr_cmd(*dev);
-		}
-		else if(!strcmp(tok, "ERROR_STATUS")) {
-			err = err_status_cmd(*dev);
-		}
-		else if(!strcmp(tok, "DCD_WRITE")) {
-		}
-		else if(!strcmp(tok, "PROMPT")) {
-		}
-		else if(!strcmp(tok, "REBOOT")) {
-		}
-		else {
-			fprintf(stderr, "Unrecognized token '%s' at line %lu\n", tok, lineno);
-			err = -1;
-		}
+	if (script_expect_opt(s, script_tok_integer, "Optional <address> value was expected") == SCRIPT_OK)
+		addr = s->token.num;
+
+	if (s->errstr)
+		return SCRIPT_ERROR;
+
+	if (script_expect_opt(s, script_tok_integer, "Optional <size> value was expected") == SCRIPT_OK)
+		size = s->token.num;
+
+	if (s->errstr)
+		return SCRIPT_ERROR;
+
+	if (get_buffer(s, type, str, &blob) < 0)
+		return SCRIPT_ERROR;
+
+	if (s->flags & SCRIPT_F_DRYRUN)
+		return SCRIPT_OK;
+
+	if (size) {
+		size = MIN(size, blob.end - blob.ptr);
 	}
+	else {
+		size = blob.end - blob.ptr;
+	}
+
+	res = SCRIPT_ERROR;
 
-	return err;
+	if (dev)
+		res = sdp_writeFile(dev, addr, blob.ptr + offset, size);
+
+	close_buffer(type, &blob);
+
+	if (res == SCRIPT_OK)
+		return SCRIPT_OK;
+
+	s->errstr = "Device not available";
+
+	return SCRIPT_ERROR;
+}
+
+static int not_implemented_cmd(script_t *s)
+{
+	s->errstr = "This function is not yet implemented.";
+
+	return SCRIPT_ERROR;
 }
 
+/*
+ * NOTE: because binary search is used, function names
+ * must be sorted in lexical order, use upper-case,
+ * the list must be terminated with a NULL element.
+ */
+script_funct_t funcs[] = {
```

Comment:
If `funcs` is unchanging, it can be declared as `static const script_funct_t funcs[]`. This will make changes to functions associated with this variable.

Reply:
Resolved, thanks.